File: 2696-the-number-of-beautiful-subsets/the-number-of-beautiful-subsets.java
class Solution {
    private int count;
    private Set<Integer> set;
    private void solve(int index, int[] nums, int k){
        if(index >= nums.length){
            count++;
            return;
        }

        //skip
        solve(index + 1, nums, k);

        //pick
        if(!set.contains(nums[index] - k)){
            set.add(nums[index]);
            solve(index + 1, nums, k);
            set.remove(nums[index]);
        }
    }

    public int beautifulSubsets(int[] nums, int k) {
        count = 0;
        set = new HashSet<>();
        Arrays.sort(nums);
        solve(0, nums, k);
        return count - 1;
    }
}

File: 2696-the-number-of-beautiful-subsets/the-number-of-beautiful-subsets.cpp
class Solution {
public:
    unordered_set<int> set;
    int count=0;
    void countSubsets(vector<int>& a, int k,int index)
    {
        if(index==a.size())
        {
            count++;
            return;
        }

        countSubsets(a,k,index+1);     // notPick

        if(!set.count(a[index]-k))
        {
            set.insert(a[index]);
            countSubsets(a,k,index+1);     // pick
            set.erase(a[index]);
        }
    }

    int beautifulSubsets(vector<int>& a, int k) 
    {
        sort(a.begin(),a.end());
        countSubsets(a,k,0);
        return count-1;    // -1 for empty subset.
    }
};